pipeline_dp/pipeline_operations.py

```diff
@@ -74,3 +86,48 @@ def sample_fixed_per_key(self, col, n: int, stage_name: str):
 
   def count_per_element(self, col, stage_name: str):
     return col | stage_name >> combiners.Count.PerElement()
+
+
+class SparkOperations(PipelineOperations):
+  """Apache Spark Core adapter."""
+
+  def map(self, rdd, fn, stage_name: str = None):
+    return rdd.map(fn)
+
+  def map_tuple(self, rdd, fn, stage_name: str = None):
+    return rdd.map(fn)
+
+  def map_values(self, rdd, fn, stage_name: str = None):
+    return rdd.mapValues(fn)
+
+  def group_by_key(self, rdd, stage_name: str = None):
+    """Group the values for each key in the RDD into a single sequence.
+
+    Args:
+      rdd: input RDD
+      stage_name: not used
+
+    Returns:
+      An RDD of tuples in which the type of the second item
+      is the pyspark.resultiterable.ResultIterable.
+
+    """
+    return rdd.groupByKey()
+
+  def filter(self, rdd, fn, stage_name: str = None):
+    return rdd.filter(fn)
+
+  def keys(self, rdd, stage_name: str = None):
+    return rdd.keys()
+
+  def values(self, rdd, stage_name: str = None):
+    return rdd.values()
+
+  def sample_fixed_per_key(self, rdd, n: int, stage_name: str = None):
+    return rdd.groupByKey()\
+      .mapValues(lambda it: list(it))\
```

Comment:
I'm thinking about case when there is a lot of elements per key. Is it possible that pyspark.resultiterable.ResultIterable represents a large lazy collection which doesn't fit in memory?
Is there some way not to materialize it to list?
For example Beam sample_per_key does in the following way:
1.There is a combiner for getting top_n elements per key
2.sample_per_key = GroupByKey().mapValues(lambda it: ((rand(), x) for x in it)).reduceValues(top_n)
i.e. using rand() to ensure a random order and choosing top n elements in this order guarantees uniformly random sampling. 

Would something like this possible in PySpark?

Reply:
Sry, in the description how Beam works for sampling, not GroupByKey but CombinePerKey is used, which means it's fine to have a large number of elements per key

Reply:
You're spot on. I hesitated implementing it this way because of potential issues in case of imbalanced distribution of values between keys. My excuse was that people dealing with huge datasets will be more likely to use the Scala/Java version of this library and Dataframe/Dataset based implementation.

However, on second thought, it's possible to prevent Out of Memory errors by using a simple implementation which I'll push with the next commit.

Spark's _top(…)_ doesn't fit well here because it would gather everything to the driver. _RDD.sampleByKey(…)_ could be used here but it required proving fractions, not _n_, so preliminary collecting of sample data to the driver (like one done in Spark's RangePartitioner) would be necessary, which is complicated.